### PR TITLE
Enable virtual libraries in dune 1.7

### PR DESCRIPTION
**Pull request by rgrinberg**
The only thing that is left to write a manual page for this feature

### CODE REVIEW

**Comment by rgrinberg** [comment]
@avsm and @samoht are invited to review the documentation as well

File: doc/variants.rst [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+The feature introduces two kinds of libraries: virtual and implementations. A
+*virtual library* corresponds to an interface (although it may contain partial
+implementation). An *implementation* of a virtual library fills in all
+unimplemented modules in the virtual library.
+
+The benefit of this partition is that other libraries may depend and compile
+against the virtual library and only select concrete implementations for these
+virtual libraries when linking executables. An example where this might be
+useful would be a virtual, cross platform, ``os`` library. This library would
```
Comment (emillon) [inline doc/variants.rst]: What do you think about making this example a bit more concrete? For example:
- a "clock" virtual library that has two implementations, "clock.unix" and "clock.win",
- a "calendar" library depending on "clock" that deals with time & date calculations,
- a "birthday-reminder" executable depending on "calendar" that computes your friends' birthdays.

Comment (rgrinberg) [reply doc/variants.rst]: Sure. But do you mean with actual code? Or just change it to describe your example?

Comment (emillon) [reply doc/variants.rst]: I was just thinking about changing the names to describe the example. Some example code under `example/` would be nice one day, but we'll probably need something simpler.

Comment (rgrinberg) [reply doc/variants.rst]: Ah ok. Fixed the examples to use clock.

**Comment by rgrinberg** [comment]
It will actually not be possible to enable this feature until external virtual libraries are properly implemented. The reason for that is that our second step closure depends on at least knowing which libraries are virtual and which libraries are implementations. This information would need to be present to even properly use virtual libraries that are installed, let alone implement them.

Oh well, writing documentation never hurts everyone is still welcome to review it. I'll just refrain from adding it to the index until the feature is ready.

**Comment by emillon** [comment]
Reminder: once this is shipped, we'll need to remove the disclaimer in the doc.

File: test/blackbox-tests/test-cases/variants/run.t [hunk 2]
```diff
@@ -1,31 +1,18 @@
...
 
   $ dune build --root invalid-virtual-lib
-  Entering directory 'invalid-virtual-lib'
-  File "dune", line 3, characters 18-21:
+  File "dune", line 3, characters 1-26:
   3 |  (virtual_modules foo bar))
-                        ^^^
-  Error: Module Foo has an implementation, it cannot be listed here
+       ^^^^^^^^^^^^^^^^^^^^^^^^^
+  Error: 'virtual_modules' is only available since version 1.7 of the dune language
```
Comment (emillon) [inline test/blackbox-tests/test-cases/variants/run.t]: I think you need to upgrade the project here

**Comment by rgrinberg** [comment]
@emillon  done

File: test/blackbox-tests/test-cases/variants/double-implementation/dune-project [hunk 3]
```diff
@@ -1,3 +1 @@
-(lang dune 1.2)
-
-(using in_development_do_not_use_variants 0.1)
\ No newline at end of file
+(lang dune 1.5)
```
Comment (emillon) [inline test/blackbox-tests/test-cases/variants/double-implementation/dune-project]:
```suggestion
(lang dune 1.7)
```

File: test/blackbox-tests/test-cases/variants/impl-private-modules/dune-project [hunk 4]
```diff
@@ -1,2 +1 @@
-(lang dune 1.2)
-(using in_development_do_not_use_variants 0.1)
\ No newline at end of file
+(lang dune 1.5)
```
Comment (emillon) [inline test/blackbox-tests/test-cases/variants/impl-private-modules/dune-project]:
```suggestion
(lang dune 1.7)
```

File: test/blackbox-tests/test-cases/variants/impl-public-modules/dune-project [hunk 5]
```diff
@@ -1,2 +1 @@
-(lang dune 1.2)
-(using in_development_do_not_use_variants 0.1)
\ No newline at end of file
+(lang dune 1.5)
```
Comment (emillon) [inline test/blackbox-tests/test-cases/variants/impl-public-modules/dune-project]:
```suggestion
(lang dune 1.7)
```

File: test/blackbox-tests/test-cases/variants/install-file/dune-project [hunk 6]
```diff
@@ -1,2 +1 @@
-(lang dune 1.2)
-(using in_development_do_not_use_variants 0.1)
\ No newline at end of file
+(lang dune 1.5)
```
Comment (emillon) [inline test/blackbox-tests/test-cases/variants/install-file/dune-project]:
```suggestion
(lang dune 1.7)
```

File: test/blackbox-tests/test-cases/variants/missing-implementation/dune-project [hunk 7]
```diff
@@ -1,3 +1 @@
-(lang dune 1.2)
-
-(using in_development_do_not_use_variants 0.1)
\ No newline at end of file
+(lang dune 1.5)
```
Comment (emillon) [inline test/blackbox-tests/test-cases/variants/missing-implementation/dune-project]:
```suggestion
(lang dune 1.7)
```

File: test/blackbox-tests/test-cases/variants/variants-simple/dune-project [hunk 8]
```diff
@@ -1,3 +1 @@
-(lang dune 1.2)
-
-(using in_development_do_not_use_variants 0.1)
\ No newline at end of file
+(lang dune 1.5)
```
Comment (emillon) [inline test/blackbox-tests/test-cases/variants/variants-simple/dune-project]:
```suggestion
(lang dune 1.7)
```

File: test/blackbox-tests/test-cases/variants/variants-sub-module/dune-project [hunk 9]
```diff
@@ -1,3 +1 @@
-(lang dune 1.2)
-
-(using in_development_do_not_use_variants 0.1)
\ No newline at end of file
+(lang dune 1.5)
```
Comment (emillon) [inline test/blackbox-tests/test-cases/variants/variants-sub-module/dune-project]:
```suggestion
(lang dune 1.7)
```

**Review by emillon — APPROVED** [review]
Thanks!

**Pull request merged**